cudax/include/cuda/experimental/__device/logical_device.cuh

```diff
@@ -0,0 +1,144 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__DEVICE_LOGICAL_DEVICE
+#define _CUDAX__DEVICE_LOGICAL_DEVICE
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/experimental/__device/all_devices.cuh>
+#include <cuda/experimental/__green_context/green_ctx.cuh>
+
+namespace cuda::experimental
+{
+namespace detail
+{
+struct logical_device_access;
+}
+
+//! @brief A non-owning representation of a CUDA device or a green context
+struct logical_device
+{
+  // Enum to indicate kind of logical device stored
+  enum class kinds
+  {
+    // Indicates logical device is a full device
+    device,
+    // Indicated logical device is a green context
+    green_context
+  };
+
+  // We might want to make this private depending on how this type ends up looking like long term,
+  // not documenting it for now
+  _CCCL_NODISCARD constexpr CUcontext context() const noexcept
+  {
+    return __ctx;
+  }
+
+  //! @brief Retrieve the device on which this logical device resides
+  _CCCL_NODISCARD constexpr device_ref underlying_device() const noexcept
```

Comment:
I am wondering whether we should just use `get_device` as that is the common spelling for getters
```suggestion
  _CCCL_NODISCARD constexpr device_ref get_device() const noexcept
```

Reply:
I will make it `get_underlying_device`